### ISSUE TITLE
Increase Performance for very large datasets

### ISSUE THREAD

**Issue by rivadblins** [RevealBi/Reveal.Sdk#325]
**Is your feature request related to a problem? Please describe.**
When we initially setup RevealBI and pointed it at our test database for MongoDB we noticed significant performance issues rendering the dashboard unusable for minutes at a time while it loaded, and the ease at which we observed the cache refreshing meant that users could cause that multi-minute load to happen without necessarily meaning to very easily.

**Describe the solution you'd like**
The Dashboard loads within seconds.

**Describe alternatives you've considered**
Instead of connecting directly to MongoDB we configured Reveal to go to a dedicated REST API backend we developed. On a first load, this API is called and runs a set of aggregation pipelines that aggregates our logs into a smaller subset of documents. These documents are persisted in a configurable collection inside MongoDB. Reveal then queries this smaller dataset to generate its charts. This still has perfomance issues on first start as it takes minutes to process the data, but after that it is not possible for the user to cause a multi-minute load again because Reveal always queries the condensed dataset.

After a first load this reduced our load times to seconds, but meant that we could not leverage the dynamic aggregation pipeline building tools in the dashboard builder since we needed to aggregate the log data prior to that. If we needed to create a new way to aggregate the data, say by user, we would need to add a new aggregation pipeline and modify our backend to support this new pipeline rather than just creating the aggregation along with the chart in the builder.

**Additional context**
- On the small end of our dataset we have 5.6 million records in one collection and closer to 86 million in another corresponding to the average logs for a customer over a period of 365 days. Only new records are ever added, and no records are ever updated.
- Load times for the dataset when using RevealBI's built in Mongo datasource was 5 minutes for the 5.6 million record database and closer to 15 minutes for the 86 million one.

**Comment by brianlagunas** [comment]
Is this the same as #288? We just release some performance improvements to the Mongo connector in our latest 1.7.1 release based on your previous feedback.

**Comment by rivadblins** [comment]
No this is not related. Even with #288 fixed the connector is still unusable for our dataset. The built in caching does not suite our needs as our data is static for any given date range and once calculated for a period never needs to be calculated for that period again. 

The current behavior I have observed while developing is that an end-user can very easily, by modifying the date range, or simply refreshing the dashboard with the refresh button, cause the cache to invalidate and a long aggregation to run again. We cannot deliver this to our customers at the ones requesting this dashboard functionality could take 10s of minutes to load and causing that when they change the date range is not feasible.

Our work around was implementing what I described above. We created a backend that runs an aggregation and then stores the results into a TTL collection before recording a metadata entry corresponding to that calculation. This stops us from reprocessing the same data again since we can check the last date we processed the data and then just process the documents that were added after that date.

The reason we are making this request is that our aggregation pipelines are now hardcoded into the backend and so our users/Product Owners will not be able to create their own relevant aggregation against the raw data that we have, severely limiting what they will be able to build.

An example we had to remove a chart that displayed the top 10 users by number of events because the two aggregations that we defined as inscope for development did not aggregate on username, and so without a backend code change we cannot build charts around the user field anymore since the individual user that a document corresponds to is lost in the aggregation process.

We want to use the existing functionality that the dashboard builder has to construct these aggregations on the fly, we just want them to persist that data. We need a solution that isn't going to reprocess that data over and over again for every user that loads the page or causes a refresh.

**Comment by rivadblins** [comment]
Even if the dashboard builder made a call to a customizable REST endpoint and passed the JSON of the aggregation pipeline to it, we could work with that as a V1 solution. We do not want to have to make our own aggregation pipeline builder just for this functionality as it already seems to exist within the dashboard builder.

We are missing out on most of the benefits of using Reveal because we had to make this custom solution to allow it to be performant to our expectations.

**Comment by MLopezIG** [comment]
So, let me see if I get it right. What you're doing is basically an ETL on your raw data, producing new collections (still using mongo) that are faster to query... however, those collections only capture some of the possible queries, so not everything that a user may want to do is possible. Correct? 

Have you evaluated doing ETL but targeting a different database? Maybe even a relational DB -with good indexes- can do better. If you're not restricted to on-prem, you might evaluate e.g. Athena/BigQuery/Redshift. Not sure if this needs to be real time in your case? 
Your ETL process might just insert the new raw data in the target database. You'd insert the raw data there, and let Reveal directly query that database (the end-user will build whatever visualization he wants). 

I'm assuming that you've hit a performance bottleneck on MongoDb, independent of Reveal, and no further improvements are to be expected from optimizing the queries (e.g. by creating indexes).

**Comment by rivadblins** [comment]
> So, let me see if I get it right. What you're doing is basically an ETL on your raw data, producing new collections (still using mongo) that are faster to query... however, those collections only capture some of the possible queries, so not everything that a user may want to do is possible. Correct?

Yes this is correct. We are pre-running data aggregations against event types and module types, which means any visualization has to work within that grouped dataset.

> Have you evaluated doing ETL but targeting a different database? Maybe even a relational DB -with good indexes- can do better. If you're not restricted to on-prem, you might evaluate e.g. Athena/BigQuery/Redshift. Not sure if this needs to be real time in your case?

We have not tried a different database architecture, but I will discuss this internally with the team if this is something we could do. This dashboard needs to be, at the very least, semi-realtime. The user should be able to open the page and see data starting from the day before and backwards without needing to wait minutes.

> Your ETL process might just insert the new raw data in the target database. You'd insert the raw data there, and let Reveal directly query that database (the end-user will build whatever visualization he wants).

The ETL process we have now works with Mongo as once the data has been aggregated we compress it down to be daily reports, so no matter how many documents there can only be 1 per day per customer. So it's not the extra collection with its processed data that is the problem, it's the raw data that we're ingesting from our logs into mongo that has a such a high volume.

> I'm assuming that you've hit a performance bottleneck on MongoDb, independent of Reveal, and no further improvements are to be expected from optimizing the queries (e.g. by creating indexes).

That is correct in the sense that we have optimized our Mongo queries and even then it is not possible to query the raw data performantly in real-time. We have a solution that we implemented, the issue is that we don't want to have to manage this new backend and these new collections as our understanding of Reveal was that it was built to abstract that away and manage the aggregation building and caching itself.

If we have to have our own backend performing the aggregations and managing the data, then we are really only using Reveal to display charts and the cost vs benefit ratio gets skewed. Myself and the team would like to continue to use Reveal, but our original concept was to use a simple charting library with our own backend, and we ended up halfway to that solution despite switching to Reveal.

Without a more hands-off approach we may end up having to drop Reveal.

**Comment by MLopezIG** [comment]
The goal of using MongoDb for real time analytics with that amount of documents sounds challenging. This is beyond caching. 

I'm wondering how far back in time do your users really need to go. In the next version of Reveal it will be possible to restrict the date filters available for querying. So I'm wondering if limiting to just a few days would be enough for most of the needs, and maybe performance would be reasonable (provided there's an index on the date field).

**Comment by rivadblins** [comment]
I discussed using a different database backend with our architect and unfortunately that is a non-stater. We have tried relational databases with this data load before and they have not performed to our expectations enough that we would switch to them from Mongo. We also have other applications that integrate with Mongo and having a separate technology stack and rewriting the existing backend to use a different database is not feasible and a non-starter.

Date filter restriction will definitely help, and we do have plans to restrict the time frame that can be queried, however for some of our largest customers even within a 60 day timeframe they will be hitting 5-10 minute load times. I also agree, that number of documents cannot be cached locally for a user it's just too much data.

The best solutions I can offer is one of the following:

1) Add some way for a visualization to pass the Aggregation Pipeline it is building to our back end, we can store that aggregation and ensure it gets run on a regular basis and the visualization can simple query a date range against a pre-aggregated data set.

This has a few drawbacks, namely that we would still need to manage that data, those pipelines, and the collections where they are stored and have to put in additional effort to get that working, making Reveal as a whole less plug-and-play that we would like, and increasing the overhead.

It also still calls into question the value proposition here, if it's our backend that's doing the heavy lifting and we need to maintain all this stuff, but reveal handles the frontend, then we are missing out on the integrated datasource functionality, your own caching solution, etc... This solution is a middle ground for us between what reveal can currently do and what our custom-made for Reveal backend can do.

2) The visualizations don't run their aggregation pipelines every single time the data needs to be refreshed, instead running the aggregation on a timed schedules and managing the data itself. 

This solution is basically "Do what we already did" or finding a similar approach and is our preferred option since we could significantly simplify our implementation and would get us more value out of Reveal than we are currently.

**Comment by MLopezIG** [comment]
@rivadblins Please note that the previous suggestion wasn't to replace MongoDb with another relational. It was about replicating your data in some other database more capable for analytics. Your ETL would just be about exporting the last relevant docs from your Mongo collection(s). Then Reveal queries this other database. As mentioned previously, if you're not restricted to on-prem, solutions like BigQuery would shine on query performance.
